src/AggregateFunctions/TimeSeries/AggregateFunctionTimeseriesChanges.h

```diff
@@ -0,0 +1,203 @@
+    using ValueType = ValueType_;
+
+    static String getName()
+    {
+        return is_resets ? "timeSeriesResetsToGrid" : "timeSeriesChangesToGrid";
+    }
+
+    struct Bucket
+    {
+        absl::flat_hash_map<TimestampType, ValueType> samples;
+
+        void add(TimestampType timestamp, ValueType value)
+        {
+            auto it = samples.find(timestamp);
+            if (it != samples.end())
+                it->second = std::max(it->second, value);
+            else
+                samples[timestamp] = value;
+        }
+
+        void merge(const Bucket & other)
+        {
+            samples.reserve(other.samples.size());
+
+            for (const auto & [timestamp, value] : other.samples)
+                add(timestamp, value);
+        }
+    };
+};
+
+template <typename Traits>
+class AggregateFunctionTimeseriesChanges final :
+    public AggregateFunctionTimeseriesBase<AggregateFunctionTimeseriesChanges<Traits>, Traits>
+{
+public:
+    static constexpr bool DateTime64Supported = true;
+
+    static constexpr bool is_resets = Traits::is_resets;
+
+    using TimestampType = typename Traits::TimestampType;
+    using IntervalType = typename Traits::IntervalType;
+    using ValueType = typename Traits::ValueType;
+
+    using Base = AggregateFunctionTimeseriesBase<AggregateFunctionTimeseriesChanges<Traits>, Traits>;
+
+    using Base::Base;
+
+    using Bucket = typename Base::Bucket;
+
+    static void serializeBucket(const Bucket & bucket, WriteBuffer & buf)
+    {
+        writeBinaryLittleEndian(bucket.samples.size(), buf);
+        for (const auto & sample : bucket.samples)
+        {
+            writeBinaryLittleEndian(sample.first, buf);
+            writeBinaryLittleEndian(sample.second, buf);
+        }
+    }
+
+    void deserializeBucket(Bucket & bucket, ReadBuffer & buf, const size_t bucket_index) const
+    {
+        size_t sample_count;
+        readBinaryLittleEndian(sample_count,buf);
+        bucket.samples.reserve(sample_count);
+
+        for (size_t s = 0; s < sample_count; ++s)
+        {
+            TimestampType timestamp;
+            readBinaryLittleEndian(timestamp, buf);
+            Base::checkTimestampInRange(timestamp, bucket_index);
+
+            ValueType value;
+            readBinaryLittleEndian(value, buf);
+
+            bucket.add(timestamp, value);
+        }
+    }
+
+private:
+    void fillResultValue(const std::deque<std::pair<TimestampType, ValueType>> & samples_in_window,
+        ValueType & result, UInt8 & null) const
+    {
+        if (samples_in_window.empty())
+        {
+            result = 0;
+            null = 1;
+            return;
+        }
+
+        UInt64 change_count = 0;
+        std::pair<TimestampType, ValueType> prev_sample = samples_in_window.front();
+        for (const auto& sample : samples_in_window)
+        {
+            if ((is_resets && sample.second < prev_sample.second) || (!is_resets && sample.second != prev_sample.second))
```

Comment:
```suggestion
            if constexpr (is_resets)
            {
                bool is_reset = (sample.second < prev_value);
                if (is_reset)
                    count++;
            }
            else
            {
                bool is_change = (sample.second != prev_value);
                if (is_change)
                    count++;
            }
```

This way we could get rid of this complicated condition. Also we don't need both parts of `previous_sample`, we only need its value.